odc/geo/cog/_s3.py

```diff
@@ -0,0 +1,318 @@
+        return sess.create_client(
+            "s3",
+            endpoint_url=self.endpoint_url,
+            aws_access_key_id=creds.access_key,
+            aws_secret_access_key=creds.secret_key,
+            aws_session_token=creds.token,
+        )
+
+    def initiate(self, **kw) -> str:
+        assert self.uploadId == ""
+        s3 = self.s3_client()
+
+        rr = s3.create_multipart_upload(Bucket=self.bucket, Key=self.key, **kw)
+        uploadId = rr["UploadId"]
+        self.uploadId = uploadId
+        return uploadId
+
+    def write_part(self, part: int, data: SomeData) -> dict[str, Any]:
+        s3 = self.s3_client()
+        assert self.uploadId != ""
+        rr = s3.upload_part(
+            PartNumber=part,
+            Body=data,
+            Bucket=self.bucket,
+            Key=self.key,
+            UploadId=self.uploadId,
+        )
+        etag = rr["ETag"]
+        return {"PartNumber": part, "ETag": etag}
+
+    @property
+    def url(self) -> str:
+        return f"s3://{self.bucket}/{self.key}"
+
+    def finalise(self, parts: list[dict[str, Any]]) -> str:
+        s3 = self.s3_client()
+        assert self.uploadId
+
+        rr = s3.complete_multipart_upload(
+            Bucket=self.bucket,
+            Key=self.key,
+            UploadId=self.uploadId,
+            MultipartUpload={"Parts": parts},
+        )
+
+        return rr["ETag"]
+
+    @property
+    def started(self) -> bool:
+        return len(self.uploadId) > 0
+
+    def cancel(self, other: str = ""):
+        uploadId = other if other else self.uploadId
+        if not uploadId:
+            return
+
+        s3 = self.s3_client()
+        if uploadId.lower() in ("all", ":all:"):
+            for uploadId in self.list_active():
+                s3.abort_multipart_upload(
+                    Bucket=self.bucket, Key=self.key, UploadId=uploadId
+                )
+            self.uploadId = ""
+        else:
+            s3.abort_multipart_upload(
+                Bucket=self.bucket, Key=self.key, UploadId=uploadId
+            )
+            if uploadId == self.uploadId:
+                self.uploadId = ""
+
+    def list_active(self):
+        s3 = self.s3_client()
+        rr = s3.list_multipart_uploads(Bucket=self.bucket, Prefix=self.key)
+        return [x["UploadId"] for x in rr.get("Uploads", [])]
+
+    def read(self, **kw):
+        s3 = self.s3_client()
+        return s3.get_object(Bucket=self.bucket, Key=self.key, **kw)["Body"].read()
+
+    def __dask_tokenize__(self):
+        return (
+            self.bucket,
+            self.key,
+            self.uploadId,
+        )
+
+    def writer(self, kw, *, client: Any = None) -> PartsWriter:
+        if client is None:
+            client = _dask_client()
+        writer = DelayedS3Writer(self, kw)
+        if client is not None:
+            writer.prep_client(client)
+        return writer
+
+    # pylint: disable=too-many-arguments
+    def upload(
```

Comment:
Does this only support Dask inputs, or non-Dask COGs too?

Reply:
no, this is specifically for Dask, it operates on Dask specific collections of "future" bytes.